Sources/SpeziOnboarding/ConsentView/ConsentDocumentExport.swift

```diff
@@ -6,42 +6,79 @@
 // SPDX-License-Identifier: MIT
 //
 
-import PDFKit
+@preconcurrency import PDFKit
+import PencilKit
+import SwiftUI
 
 /// A type representing an exported `ConsentDocument`. It holds the exported `PDFDocument` and the corresponding document identifier String.
-public actor ConsentDocumentExport {
+@Observable
+public final class ConsentDocumentExport: Equatable, Sendable {
     /// Provides default values for fields related to the `ConsentDocumentExport`.
     public enum Defaults {
         /// Default value for a document identifier.
         /// This identifier will be used as default value if no identifier is provided.
         public static let documentIdentifier = "ConsentDocument"
     }
 
-    private var cachedPDF: PDFDocument
-    
+    let asyncMarkdown: () async -> Data
+    let exportConfiguration: ConsentDocument.ExportConfiguration
+    var cachedPDF: PDFDocument?
+
     /// An unique identifier for the exported `ConsentDocument`.
     /// Corresponds to the identifier which was passed  when creating the `ConsentDocument` using an `OnboardingConsentView`.
     public let documentIdentifier: String
+    
+    /// The name of the person which signed the document.
+    public var name = PersonNameComponents()
+    #if !os(macOS)
+    /// The signature of the signee as drawing.
+    public var signature = PKDrawing()
+    /// The image generated from the signature drawing.
+    public var signatureImage = UIImage()
+    #else
+    /// The signature of the signee as string.
+    public var signature = String()
+    #endif
+
     /// The `PDFDocument` exported from a `ConsentDocument`.
     /// This property is asynchronous and accesing it potentially triggers the export of the PDF from the underlying `ConsentDocument`,
     /// if the `ConsentDocument` has not been previously exported or the `PDFDocument` was not cached.
     /// For now, we always require a PDF to be cached to create a ConsentDocumentExport. In the future, we might change this to lazy-PDF loading.
-    public var pdf: PDFDocument {
+    @MainActor public var pdf: PDFDocument {
         get async {
-            cachedPDF
+            if let cached = cachedPDF {
+                return cached
+            } else {
+                cachedPDF = await export()
+                // If the export failed, return an empty document.
+                return cachedPDF ?? .init()
+            }
```

Comment:
That's a good point, however the problem is that the PDF export could actually fail. When generating the PDFDocument from TPPDF, the generator could throw, and/or the data could be nil (see [here](https://github.com/RealLast/SpeziOnboarding/blob/PDFPagination/Sources/SpeziOnboarding/ConsentView/ConsentDocumentExport%2BExport.swift#L175) lines 175-183.

I am not sure under which circumstances TPPDF would throw, or in what cases the data might be incompatible with the PDFDocument, but it could happen. As an alternative, we would need to make the export function throwing. But then, also the async pdf variable of ConsentDocumentExport would still need to return an empty document in that case (or simply be throwing as well?)

Reply:
Yes, I think the best option is to bubble the issue up to the `pdf` variable. We can also make this property throwing and pass the error up to the caller to handle it there. Better than giving back an empty PDF.

Reply:
Hi @RealLast, just got a message about your comment (seems like I am still subscribed to this PR), so I'll give my two cents here:

TPPDF is designed to throw errors in the case when the APIs are not used correctly (i.e. when accessing indicies of a table which are out of bounds).

You can actually lookup the [throwing code](https://github.com/search?q=repo%3Atechprimate%2FTPPDF+path%3A%2F**%2F*.swift+%22throw+%22&type=code) in the repository to see it yourself.

Hope this helps.

Reply:
Perfect, I have made the export throwing and propagate a potential exception up to the ``pdf`` variable.

Reply:
If we can be sure that the PDF was constructed properly, can we avoid to make the pdf property throwing (just using a precondition there)? If we propagate the error, I’m not quite sure it the caller really knows what to do with the error except to ignore it. Looking at the linked „throwing code“, it seems like an error could only occur if we mess up in the library itself? Not an expert on the topic though.

Reply:
The problem is that we probably do not really have full "control" over the content of the PDF file, as the content is generated based on a markdown file. Theoretically, users of Spezi could write down whatever they like in that markdown file. I am not sure if we can guarantee that this content is always "valid". For example, if we add support for tables to markdown in the future, can we guarantee that this table would not be empty (i.e.,  PDFError.tableIsEmpty)? Also, even if TPPDF does not throw an error, there is still the chance that the conversion from ``TPPDF.PDFDocument`` to ``PDFKit.PDFDocument`` fails, if PDFKit cannot parse the data (see [here](https://github.com/RealLast/SpeziOnboarding/blob/PDFPagination/Sources/SpeziOnboarding/ConsentView/ConsentDocumentExport%2BExport.swift#L178)). Not sure if that would ever really happen though.

Reply:
Thanks for the explanation 🚀 

Is my understanding correct, that the pdf generation fails if (and only if) the original markdown document contains  invalid elements (that cannot converted to a pdf document). Or phrased differently, if PDF generation fails it fails all the time or if it doesn't fail it will never fail. If that would be the case, we could verify that a markdown document can generate a valid PDF document at startup (or when we receive the markdown document).

In any case, would it be feasible to do verification of the pdf generation of the markdown file on startup (or once we receive that markdown file) anyways? Just to ensure that such errors can be caught as early as possible.

Just adding some of my general thoughts here. I really do not have any deep insights of the development or what you already discussed with @PSchmiedmayer. So well free to ignore whatever I said if it doesn't apply or was already though of 🙈

Reply:
Good point, thanks for the considerations! You are right, I have the same understanding. The PDF generation should only fail if we provide invalid content, and then it should fail consistently, either because TPPDF throws or the generated data is invalid. Given that the markdown content is static and does not change, we could definitely test this at startup. However, then the "lazy PDF export" probably would not thaat much sense anymore (as we already test the PDF generation during startup).

Reply:
Thanks @PSchmiedmayer! I have changed it accordingly and made it non-throwing. Yet, the async functionality is still available, and if no cachedPDF was provided, then the export is triggered when acessing the PDF. We agreed to verify the content beforehand, so this should never throw, but still I would need to return a default value if it does. Thus, just for completeness, I added this statement:
  ```
guard let pdf = try? await export() else {
                return .init()
            }
```
As long as always pass known good markdown text to the export function, this should never get called.